Observer/SalesQuoteProductAddAfter.php

```diff
@@ -0,0 +1,157 @@
+use Magento\Quote\Model\Quote;
+use Magento\Catalog\Model\CategoryFactory;
+use Magento\Checkout\Model\Session;
+
+class SalesQuoteProductAddAfter implements ObserverInterface
+{
+    /**
+     * Klaviyo Cart Search Interface
+     * @var  CartSearchRepository $_cartSearchRepository
+     */
+    protected $_cartSearchRepository;
+
+    /**
+     * Magento Category Factory
+     * @var CategoryFactory $_categoryFactory
+     */
+    protected $_categoryFactory;
+
+    /**
+     * Magento Checkout Session data
+     * @var Session $_checkoutSession
+     */
+    protected $_checkoutSession;
+
+    /**
+     * SalesQuoteProductAddAfter constructor.
+     * @param CartSearchRepository $cartSearchRepository
+     * @param CategoryFactory $categoryFactory
+     * @param Session $checkoutsession
+     */
+    public function __construct(
+        CartSearchRepository $cartSearchRepository,
+        CategoryFactory $categoryFactory,
+        Session $checkoutsession
+    )
+    {
+        $this->_cartSearchRepository = $cartSearchRepository;
+        $this->_categoryFactory = $categoryFactory;
+        $this->_checkoutSession = $checkoutsession;
+    }
+
+    public function execute(Observer $observer)
+    {
+        $quote = $observer->getData('items')[0]->getQuote();
+        $addedItems = $observer->getData('items');
+
+        foreach ( $addedItems as $item ){
+            $this->klAddedToCartItemData( $quote, $item );
+        }
+    }
+
+    /**
+     * Build Added to Cart payload for current Added item
+     * @param $quote
+     * @param $addedItem
+     */
+    public function klAddedToCartItemData( $quote, $addedItem )
+    {
+        $addedProduct = $addedItem->getProduct();
+        $addedItemData = array(
+            'AddedItemCategories' => (array) $this->getCategoryName( $addedProduct->getCategoryIds() ),
+            'AddedItemDescription' => (string) strip_tags( $addedProduct->getDescription() ),
+            'AddedItemImageUrlKey' => (string) $addedProduct->getData('small_image'),
+            'AddedItemPrice' => (float) $addedProduct->get_price(),
+            'AddedItemQuantity' => (int) $addedItem->getQty(),
+            'AddedItemProductID' => (int) $addedProduct->getId(),
+            'AddedItemProductName' => (string) $addedItem->getName(),
+            'AddedItemSku' => (string) $addedProduct->getSku(),
+            'AddedItemUrl' => (string) $addedProduct->getProductUrl()
+        );
+
+        $klAddedToCartPayload = array_merge( $this->klBuildCartData( $quote, $addedItem ), $addedItemData );
+
+        $this->getCheckoutSession()->setKlAddedToCartKey( $klAddedToCartPayload );
+    }
+
+    /**
+     * Helper function to add items from cart to Added to Cart payload
+     * @param $quote
+     * @param $addedItem
+     * @return array
+     */
+    public function klBuildCartData( $quote, $addedItem )
+    {
+        $cartItems = $quote->getItems() ? $quote->getitems() : array();
```

Comment:
can you elaborate on the logic here?

Reply:
`getItems` can return null in which case we want to instead return an empty array. you can use a null coalescing operator (php 7+) here instead if you want to clean it up a bit further
```suggestion
        $cartItems = $quote->getItems() ?? array();
```

Reply:
looks like it'll set `$cartItems` to what `getItems()` returns, or an empty array if `getItems()` doesn't return anything